pkg/pipeline/transform/transform_anomaly.go

```diff
@@ -0,0 +1,243 @@
+	}
+	alpha := anomalyConfig.EWMAAlpha
+	if alpha <= 0 {
+		alpha = 2.0 / (float64(window) + 1.0)
+	}
+	if len(anomalyConfig.KeyFields) == 0 {
+		anomalyConfig.KeyFields = []string{"SrcAddr", "DstAddr", "Proto"}
+	}
+	if anomalyConfig.Algorithm == "" {
+		anomalyConfig.Algorithm = api.AnomalyAlgorithmZScore
+	}
+
+	anomalyLog.Infof("NewTransformAnomaly algorithm=%s window=%d baselineWindow=%d", anomalyConfig.Algorithm, window, baselineWindow)
+	return &Anomaly{
+		states:         make(map[string]*anomalyState),
+		config:         anomalyConfig,
+		windowSize:     window,
+		baselineWindow: baselineWindow,
+		sensitivity:    sensitivity,
+		alpha:          alpha,
+		opMetrics:      opMetrics,
+	}, nil
+}
+
+// Transform calculates anomaly scores per key and appends anomaly fields.
+func (a *Anomaly) Transform(entry config.GenericMap) (config.GenericMap, bool) {
+	value, err := utils.ConvertToFloat64(entry[a.config.ValueField])
+	if err != nil {
+		anomalyLog.Errorf("unable to convert %s to float: %v", a.config.ValueField, err)
+		return entry, false
+	}
+	key := a.buildKey(entry)
+
+	a.mu.Lock()
+	state, ok := a.states[key]
+	if !ok {
+		state = &anomalyState{}
+		a.states[key] = state
+	}
+	anomalyType, score := a.score(state, value)
+	state.addValue(value, a.windowSize)
+	stateSize := len(state.values)
+	a.mu.Unlock()
+
+	output := entry.Copy()
+	output["anomaly_score"] = score
+	output["anomaly_type"] = anomalyType
+	output["baseline_window"] = stateSize
+
+	return output, true
+}
+
+func (a *Anomaly) score(state *anomalyState, value float64) (string, float64) {
+	if len(state.values) < a.baselineWindow {
+		if !state.initialized {
+			state.baseline = value
+			state.initialized = true
+		}
+		return "warming_up", 0
+	}
+
+	switch a.config.Algorithm {
+	case api.AnomalyAlgorithmEWMA:
+		return a.scoreEWMA(state, value)
+	case api.AnomalyAlgorithmZScore:
+		fallthrough
+	default:
+		return a.scoreZScore(state, value)
+	}
+}
+
+func (a *Anomaly) scoreEWMA(state *anomalyState, value float64) (string, float64) {
+	if !state.initialized {
+		state.baseline = value
+		state.initialized = true
+	}
+	deviation := value - state.baseline
+	stddev := state.stddev()
+	if stddev == 0 {
+		stddev = math.Max(math.Abs(state.baseline)*1e-6, 1e-9)
+	}
+	score := math.Abs(deviation) / stddev
+	state.baseline = state.baseline + a.alpha*(value-state.baseline)
+	anomalyType := "normal"
+	if score >= a.sensitivity {
+		if deviation > 0 {
+			anomalyType = "ewma_high"
+		} else {
+			anomalyType = "ewma_low"
+		}
+	}
+	return anomalyType, score
+}
+
+func (a *Anomaly) scoreZScore(state *anomalyState, value float64) (string, float64) {
+	mean := state.mean()
+	stddev := state.stddev()
+	if stddev == 0 {
+		stddev = math.Max(math.Abs(mean)*1e-6, 1e-9)
+	}
+	score := math.Abs(value-mean) / stddev
+	anomalyType := "normal"
+	if score >= a.sensitivity {
+		if value > mean {
+			anomalyType = "zscore_high"
+		} else {
+			anomalyType = "zscore_low"
+		}
+	}
```

Comment:
Should we create an enum for that ?

You could then use it as return type instead of string